### PR TITLE
Fix data race in actors

### LINKED ISSUE #2967: Data race of actor
## In what area(s)?

> /area actor

## What version of Dapr?

> 1.0.x

## Expected Behavior

Codes can handle concurrency robustly.

## Actual Behavior

Tether are some data race  problems in implementation of `actor` and `actorsRuntime`. And there are some errors when we run test with race detect.

Related issue: #2961 

### actor.disposed
Read:
>https://github.com/dapr/dapr/blob/034022589c6fb4f9cf32aebc6278bd18d45e23b5/pkg/actors/actor.go#L63

Write:
>https://github.com/dapr/dapr/blob/034022589c6fb4f9cf32aebc6278bd18d45e23b5/pkg/actors/actor.go#L93

### actor.disposeCh
Read:
>https://github.com/dapr/dapr/blob/034022589c6fb4f9cf32aebc6278bd18d45e23b5/pkg/actors/actor.go#L92

Write:
>https://github.com/dapr/dapr/blob/034022589c6fb4f9cf32aebc6278bd18d45e23b5/pkg/actors/actor.go#L69

### actorsRuntime.appHealthy
Read:
>https://github.com/dapr/dapr/blob/034022589c6fb4f9cf32aebc6278bd18d45e23b5/pkg/actors/actors.go#L155

Write:
>https://github.com/dapr/dapr/blob/034022589c6fb4f9cf32aebc6278bd18d45e23b5/pkg/actors/actors.go#L189

## Steps to Reproduce the Problem

```Shell
cd pkg/actors
go test . -race
```
## Release Note

RELEASE NOTE: **FIX** Data race of actor.

### PR DESCRIPTION
**Pull request by luckyxiaoqiang**
# Description

Fix data race in actors.

## Issue reference

Close #2967 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Review by halspang — COMMENTED** [review]
Change looks good overall. Can you bring it up to date with master?

**Comment by luckyxiaoqiang** [comment]
> Change looks good overall. Can you bring it up to date with master?

Updated.

**Review by halspang — APPROVED** [review]

File: pkg/actors/actor.go [hunk 1]
```diff
@@ -93,10 +107,12 @@ func (a *actor) lock(reentrancyID *string) error {
 func (a *actor) unlock() {
 	pending := a.pendingActorCalls.Dec()
 	if pending == 0 {
+		a.disposeLock.Lock()
 		if !a.disposed && a.disposeCh != nil {
 			a.disposed = true
 			close(a.disposeCh)
 		}
+		a.disposeLock.Unlock()
```
Comment (skyao) [inline pkg/actors/actor.go:115]: Suggest to use defer here for unlock(), since close() method is not safe enough (if close() panic, we can't unlock).

**Review by skyao — CHANGES_REQUESTED** [review]

File: pkg/actors/actor.go (hunk 1 above)
Comment (luckyxiaoqiang) [reply pkg/actors/actor.go]: Updated. Please review again.

**Review by artursouza — APPROVED** [review]

**Review by skyao — APPROVED** [review]

**Comment by skyao** [comment]
@luckyxiaoqiang thank you for PR.

**Pull request merged**